### PR TITLE
Remove duplicates in rustdoc

### LINKED ISSUE #43934: rustdoc: Duplicated trait entries if re-exported from multiple modules
If you generate the docs of
  https://github.com/sdroege/gstreamer-rs/tree/50a1535771d711b419eee499fb6b7c48d72edd86
there will be multiple entries linking to the same traits on the main page of the docs.

Reason here is that the traits are re-exported in lib.rs from `auto::*` and `auto::traits::*`. Removing the latter fixes it and has no negative side effects in this specific case, but having the entries there twice was confusing nonetheless.

@GuillaumeGomez

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #43934.

Two things however:

 1. I'm not happy with the current check. It seems completely overkill and unsatisfying.
 2. I have no idea how to test if there is only one element and not two.

r? @rust-lang/docs

### CODE REVIEW

**Comment by ollie27** [comment]
This really needs to be done when inlining instead. We need to not inline things from globs if they are shadowed.

> I have no idea how to test if there is only one element and not two.

You can use `@count` in rustdoc tests.
https://github.com/rust-lang/rust/blob/80be2f8697e869e75b0b341b0c1eecef694ee5b9/src/etc/htmldocck.py#L99-L100

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -1764,6 +1764,18 @@ fn item_module(w: &mut fmt::Formatter, cx: &Context,
     }
 
     indices.sort_by(|&i1, &i2| cmp(&items[i1], &items[i2], i1, i2));
+    indices.dedup_by_key(|i| (items[*i].def_id,
+                              if items[*i].name.as_ref().is_some() {
+                                  Some(full_path(cx, &items[*i]).clone())
+                              } else {
+                                  items[*i].name.clone()
+                              },
```
Comment (QuietMisdreavus) [inline src/librustdoc/html/render.rs]: I don't see how this else branch will return anything but None? Also was there any major reason that `is_some` needed the `as_ref` there?

Comment (GuillaumeGomez) [inline src/librustdoc/html/render.rs]: To avoid a panic mainly. But I could directly return `None`, indeed.

Comment (frewsxcv) [inline src/librustdoc/html/render.rs]: does something like `items.get(*i).map(|n| full_path(cx, &items[*i]))` work here?

**Comment by GuillaumeGomez** [comment]
@ollie27: Nice, thanks!

**Comment by GuillaumeGomez** [comment]
Fixed the tidy issue (deprecating TODO is strange).

**Comment by QuietMisdreavus** [comment]
I'm willing to land this as-is. If we want to change how this deduplication happens later, we only need to take out one statement to revert this change. (Plus, we have a test for this now. `:3`)

@bors r+ rollup

**Pull request merged**